src/ModuleBuilder.hs

```
@@ -87,19 +93,29 @@ execModuleBuilderT s m = snd <$> runModuleBuilderT s m
 emitDefn :: MonadModuleBuilder m => Definition -> m ()
 emitDefn def = liftModuleState $ modify $ \s -> s { builderDefs = builderDefs s `snoc` def }
 
+-- | A parameter name suggestion
+data ParameterName
+  = NoParameterName
+  | ParameterName ShortByteString
+  deriving (Eq, Ord, Read, Show, Typeable, Data, Generic)
+
+instance IsString ParameterName where
+  fromString = ParameterName . fromString
```

The `IsString` instance of `ShortByteString` will silently truncate non-ASCII characters. In `llvm-hs-pure` we thus have added an [explicit check](https://github.com/llvm-hs/llvm-hs/blob/llvm-5/llvm-hs-pure/src/LLVM/AST/Name.hs#L35) and throw an error rather than silently throwing data away. It seems like a good idea to be consistent and apply this here as well.

---

Consistency is good, but what I don't quite understand is why the `fromString` way of constructing these kinds of names should have this check and not the others?

---

I’m not sure what you’re referring to here by “others”. If you don’t use the `IsString` instance, you need to supply a `ShortByteString` and at that point the conversion has already happened so it is up to you to decide how you want to handle this.

---

The others are (in the case of `Name`) the `Name` constructor, as you're implying, and as an example `Name (fromString nonAscii)` won't give an error but `fromString nonAscii` will.

There exist solutions that are consistent also between the different ways of constructing names (i.e. `fromString` and the `Name` constructor have the same checks). Here are some ideas:

* Perform no checks: just let LLVM or others that can't handle non-ASCII handle the problem (probably not very nice to debug)
* Perform checks at the LLVM interop-layer (consistent, but errors come late)
* Use an abstract type `AsciiString` which is a `ShortByteString` internally but has additional checks at construction time

---

LLVM doesn’t care about the encoding it just interprets it as a sequence of bytes (not sure how well it deals if you try to print such IR but I think it’s reasonably good at escaping random bytes), that’s why we use a `ShortByteString`` rather than `String` or `Text`. So supplying random bytes that are not ASCII is not the problem here, the problem is converting from `String` to `ShortByteString` since the `IsString` instance of `ShortByteString` will silently truncate bytes.

Performing checks at the LLVM interop-layer thus doesn’t make sense since LLVM is not the problem here. A newtype would work but it’s also unnecessarily restrictive. There is no reason to restrict users here: If someone wants to use UTF-8 (or some other encoding) encoded names that’s perfectly fine, they just need to handle the conversion to `ShortByteString` themselves. I guess we could use a newtype and require a separate import to get the constructor but that seems like too much trouble for what is just intended as a convenient check to get slightly better erro checking.

---

Ah, I wasn't aware that LLVM actually could handle non-ASCII (their language reference docs are somewhat vague about it). Then the current solution seems more reasonable.

From my perspective and use of the library it seems like it'd be equally reasonable to use UTF-8 in `fromString`. Have you considered doing that?

---

The last time this came up, I decided to go for the current solution since adding UTF-8 would require an extra dependency and nobody complained about not having Unicode support. But if there’s demand for it, I’m open to adding a dependency on `utf8-string` and use `UTF-8`.